File: app/src/main/java/com/originb/inkwisenote2/modules/noterelation/worker/NoteRelationWorker.java
package com.originb.inkwisenote2.modules.noterelation.worker;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import androidx.annotation.NonNull;
import androidx.work.Worker;
import androidx.work.WorkerParameters;
import com.google.android.gms.common.util.CollectionUtils;
import com.originb.inkwisenote2.common.Logger;
import com.originb.inkwisenote2.config.AppState;
import com.originb.inkwisenote2.modules.backgroundjobs.Events;
import com.originb.inkwisenote2.modules.noterelation.data.NoteRelation;
import com.originb.inkwisenote2.modules.noterelation.data.NoteRelationDao;
import com.originb.inkwisenote2.modules.noterelation.data.TextProcessingStage;
import com.originb.inkwisenote2.modules.noterelation.service.NoteTfIdfLogic;
import com.originb.inkwisenote2.modules.ocr.data.NoteTermFrequencyDao;
import com.originb.inkwisenote2.modules.smartnotes.data.SmartBookPagesDao;
import com.originb.inkwisenote2.modules.smartnotes.data.AtomicNoteEntity;
import com.originb.inkwisenote2.modules.smartnotes.data.SmartBookPage;
import com.originb.inkwisenote2.modules.ocr.data.NoteTermFrequency;
import com.originb.inkwisenote2.common.ListUtils;
import com.originb.inkwisenote2.functionalUtils.Try;
import com.originb.inkwisenote2.modules.repositories.Repositories;
import com.originb.inkwisenote2.modules.repositories.SmartNotebook;
import com.originb.inkwisenote2.modules.repositories.SmartNotebookRepository;
import org.greenrobot.eventbus.EventBus;
import org.jetbrains.annotations.NotNull;

import java.util.*;
import java.util.stream.Collectors;

public class NoteRelationWorker extends Worker {
    private final NoteTfIdfLogic noteTfIdfLogic;
    private final NoteTermFrequencyDao noteTermFrequencyDao;
    private final NoteRelationDao noteRelationDao;
    private final SmartNotebookRepository smartNotebookRepository;
    private final SmartBookPagesDao smartBookPagesDao;

    private final Handler mainHandler;

    private final Integer TF_IDF_RELATION = 1;

    private final Logger logger = new Logger("NoteRelationWorker");

    public NoteRelationWorker(@NonNull @NotNull Context context, @NonNull @NotNull WorkerParameters workerParams) {
        super(context, workerParams);
        this.smartNotebookRepository = Repositories.getInstance().getSmartNotebookRepository();
        this.smartBookPagesDao = Repositories.getInstance().getNotesDb().smartBookPagesDao();

        noteTfIdfLogic = new NoteTfIdfLogic(Repositories.getInstance());
        noteTermFrequencyDao = Repositories.getInstance().getNotesDb().noteTermFrequencyDao();

        noteRelationDao = Repositories.getInstance().getNotesDb().noteRelationDao();

        mainHandler = new Handler(Looper.getMainLooper());
    }


    @NotNull
    @Override
    public Result doWork() {
        Try.to(() -> getInputData().getLong("book_id", -1), logger).get()
                .ifPresent(this::findRelatedNotesOfSmartBook);

        return Result.success();
    }

    public void findRelatedNotesOfSmartBook(long bookId) {
        logger.debug("Find related notes for bookId (new flow): " + bookId);

        Optional<SmartNotebook> smartBookOpt = smartNotebookRepository.getSmartNotebooks(bookId);
        if (!smartBookOpt.isPresent()) {
            logger.error("SmartNotebook doesn't exists for bookId: " + bookId);
            return;
        }

        SmartNotebook smartNotebook = smartBookOpt.get();
        logger.debug("Notebook bookId: " + bookId + " contains number of notes: " + smartNotebook.getAtomicNotes().size());
        for (AtomicNoteEntity atomicNote : smartNotebook.getAtomicNotes()) {
            findRelatedNotes(smartNotebook.getSmartBook().getBookId(), atomicNote);
        }

        EventBus.getDefault().post(new Events.NoteStatus(smartNotebook, TextProcessingStage.NOTE_READY));
    }

    public void findRelatedNotes(long bookId, AtomicNoteEntity noteEntity) {

        Set<Long> relatedNoteIds = getNoteIdsRelatedByTfIdf(noteEntity.getNoteId());
        relatedNoteIds.remove(noteEntity.getNoteId());
        if (relatedNoteIds.isEmpty()) {
            noteRelationDao.deleteByNoteId(noteEntity.getNoteId());
            return;
        }

        Map<Long, Long> noteToBookMap = smartBookPagesDao.getSmartBookPagesOfNote(relatedNoteIds).stream()
                .collect((Collectors.toMap(SmartBookPage::getNoteId, SmartBookPage::getBookId)));

        logger.debug("Related noteIds of bookId: " + bookId, ListUtils.listOf(noteEntity, relatedNoteIds));

        Set<NoteRelation> noteRelations = relatedNoteIds.stream()
                .filter(noteId -> noteId > 0)
                .filter(noteToBookMap::containsKey)
                .map(relatedNoteId ->
                        new NoteRelation(noteEntity.getNoteId(),
                                relatedNoteId,
                                bookId,
                                noteToBookMap.get(relatedNoteId),
                                TF_IDF_RELATION)
                )
                .collect(Collectors.toSet());

        noteRelationDao.deleteByNoteId(noteRelations.stream()
                .map(NoteRelation::getNoteId)
                .collect(Collectors.toList()));

        noteRelationDao.deleteByNoteId(noteRelations.stream()
                .map(NoteRelation::getRelatedNoteId)
                .collect(Collectors.toList()));

        noteRelationDao.insertNoteRelatedNotes(noteRelations);

        if (CollectionUtils.isEmpty(noteRelations)) return;

        mainHandler.post(() -> {
            // Code to be executed on the main thread
            AppState.updatedRelatedNotes(noteRelations);
        });

    }

    private Set<Long> getNoteIdsRelatedByTfIdf(long noteId) {
        Map<String, Double> tfIdfScores = noteTfIdfLogic.getTfIdf(noteId);
        Set<String> filteredTerms = new HashSet<>();
        for (String key : tfIdfScores.keySet()) {
            if (tfIdfScores.get(key) > 0.1) {
                filteredTerms.add(key);
            }
        }

        Set<Long> relatedNoteIds = noteTermFrequencyDao.getNoteIdsForTerms(filteredTerms)
                .stream().map(NoteTermFrequency::getNoteId)
                .collect(Collectors.toSet());

        return relatedNoteIds;
    }
}

File: app/src/main/java/com/originb/inkwisenote2/modules/noterelation/worker/NoteRelationWorker.kt
package com.originb.inkwisenote2.modules.noterelation.worker

import android.content.Context
import android.os.Handler
import android.os.Looper
import androidx.work.Worker
import androidx.work.WorkerParameters
import com.google.android.gms.common.util.CollectionUtils
import com.originb.inkwisenote2.common.Logger
import com.originb.inkwisenote2.config.AppState
import com.originb.inkwisenote2.functionalUtils.Try
import com.originb.inkwisenote2.modules.backgroundjobs.Events.NoteStatus
import com.originb.inkwisenote2.modules.noterelation.data.NoteRelation
import com.originb.inkwisenote2.modules.noterelation.data.NoteRelationDao
import com.originb.inkwisenote2.modules.noterelation.data.TextProcessingStage
import com.originb.inkwisenote2.modules.noterelation.service.NoteTfIdfLogic
import com.originb.inkwisenote2.modules.ocr.data.NoteTermFrequencyDao
import com.originb.inkwisenote2.modules.repositories.AtomicNotesDomain
import com.originb.inkwisenote2.modules.repositories.SmartNotebookRepository
import com.originb.inkwisenote2.modules.smartnotes.data.AtomicNoteEntity
import com.originb.inkwisenote2.modules.smartnotes.data.SmartBookPagesDao
import org.greenrobot.eventbus.EventBus
import java.util.*
import java.util.concurrent.Callable

class NoteRelationWorker(
    context: Context,
    workerParams: WorkerParameters,
    private val noteTfIdfLogic: NoteTfIdfLogic,
    private val noteTermFrequencyDao: NoteTermFrequencyDao,
    private val noteRelationDao: NoteRelationDao,
    private val smartNotebookRepository: SmartNotebookRepository,
    private val smartBookPagesDao: SmartBookPagesDao,
    private val atomicNotesDomain: AtomicNotesDomain
) : Worker(context, workerParams) {
    private val mainHandler: Handler

    private val TF_IDF_RELATION = 1

    private val logger = Logger("NoteRelationWorker")

    init {
        this.mainHandler = Handler(Looper.getMainLooper())
    }


    private data class RelationInput(var bookId: Long = 0, var noteId: Long = 0)

    override fun doWork(): Result {
        val result = Try.to(
            Callable {
                val bookId = getInputData().getLong("book_id", -1)
                val noteId = getInputData().getLong("note_id", -1)
                RelationInput(bookId, noteId)
            },
            logger
        ).get()
        Optional.ofNullable(result)
            .filter { it.bookId != -1L && it.noteId != -1L }
            .ifPresent { findRelatedNotesOfSmartBook(it) }
        return Result.success()
    }

    private fun findRelatedNotesOfSmartBook(relationInput: RelationInput) {
        val bookId = relationInput.bookId
        val noteId = relationInput.noteId
        logger.debug("Find related notes for bookId (new flow): $bookId")

        if (smartNotebookRepository.getSmartNotebooks(bookId) == null) {
            logger.error("SmartNotebook doesn't exists for bookId: $bookId")
            return
        }

        val atomicNote = atomicNotesDomain.getAtomicNote(noteId)
        findRelatedNotes(bookId, atomicNote)
        EventBus.getDefault().post(NoteStatus(bookId, TextProcessingStage.NOTE_READY))
    }

    fun findRelatedNotes(bookId: Long, noteEntity: AtomicNoteEntity) {
        val relatedNoteIds = getNoteIdsRelatedByTfIdf(noteEntity.noteId)
        relatedNoteIds.remove(noteEntity.noteId)
        if (relatedNoteIds.isEmpty()) {
            noteRelationDao.deleteByNoteId(noteEntity.noteId)
            return
        }

        val pages = smartBookPagesDao.getSmartBookPagesOfNote(relatedNoteIds)
        val noteToBookMap: Map<Long, Long> = pages.associate { it.noteId to it.bookId }

        logger.debug("Related noteIds of bookId: $bookId", listOf(noteEntity, relatedNoteIds))

        val noteRelations = relatedNoteIds
            .filter { it > 0L && noteToBookMap.containsKey(it) }
            .mapNotNull { relatedNoteId ->
                val relatedBookId = noteToBookMap[relatedNoteId] ?: return@mapNotNull null
                NoteRelation(
                    noteEntity.noteId,
                    relatedNoteId,
                    bookId,
                    relatedBookId,
                    TF_IDF_RELATION
                )
            }
            .toMutableSet()

        val noteIdsToClear = mutableListOf(noteEntity.noteId)
        noteIdsToClear.addAll(relatedNoteIds)
        noteRelationDao.deleteByNoteId(noteIdsToClear)
        noteRelationDao.insertNoteRelatedNotes(noteRelations)

        if (CollectionUtils.isEmpty(noteRelations)) return

        mainHandler.post {
            AppState.updatedRelatedNotes(noteRelations)
        }
    }

    private fun getNoteIdsRelatedByTfIdf(noteId: Long): MutableSet<Long> {
        val tfIdfScores = noteTfIdfLogic.getTfIdf(noteId)
        val filteredTerms = tfIdfScores.keys.filter { (tfIdfScores[it] ?: 0.0) > 0.1 }.toMutableSet()

        return noteTermFrequencyDao.getNoteIdsForTerms(filteredTerms)
            .map { it.noteId }
            .toMutableSet()
    }
}
